pkg/kubelet/kubelet_pods.go

```diff
@@ -1831,6 +1838,57 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			oldStatusPtr = &oldStatus
 		}
 		status := convertContainerStatus(cStatus, oldStatusPtr)
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			if status.State.Running != nil {
+				var requests, limits v1.ResourceList
+				// oldStatus should always exist if container is running
+				oldStatus, oldStatusFound := oldStatuses[cName]
+				// Set initial limits from container's spec upon transition to Running state
+				// For cpu & memory, limits queried from runtime via CRI always supercedes spec.limit
+				// For ephemeral-storage, a running container's status.limit equals spec.limit
+				determineResource := func(rName v1.ResourceName, ctrStsRes, ctrRes, oldStsRes, resource v1.ResourceList) {
+					if res, found := ctrStsRes[rName]; found {
+						resource[rName] = res.DeepCopy()
+						return
+					}
+					if oldStatusFound {
+						if oldStatus.State.Running == nil || status.ContainerID != oldStatus.ContainerID {
+							if res, exists := ctrRes[rName]; exists {
+								resource[rName] = res.DeepCopy()
+							}
+						} else {
+							if oldStsRes != nil {
+								if res, exists := oldStsRes[rName]; exists {
+									resource[rName] = res.DeepCopy()
+								}
+							}
+						}
+					}
+				}
+				if containerResources[cName].Limits != nil {
+					limits = make(v1.ResourceList)
+					determineResource(v1.ResourceCPU, cStatus.Resources.Limits, containerResources[cName].Limits, oldStatus.Resources.Limits, limits)
+					determineResource(v1.ResourceMemory, cStatus.Resources.Limits, containerResources[cName].Limits, oldStatus.Resources.Limits, limits)
+					if ephemeralStorage, found := containerResources[cName].Limits[v1.ResourceEphemeralStorage]; found {
+						limits[v1.ResourceEphemeralStorage] = ephemeralStorage.DeepCopy()
+					}
+				}
+				if containerResources[cName].Allocations != nil {
+					requests = make(v1.ResourceList)
+					determineResource(v1.ResourceCPU, cStatus.Resources.Requests, containerResources[cName].Allocations, oldStatus.Resources.Requests, requests)
```

Comment:
Why the behavior difference between cpu and memory? For my education.

Reply:
Today, cgroup supports Requests (cpu.shares) and Limits (cpu.cfs_quota_us) for CPU but not for memory which just supports limits (memory.limit_in_bytes)

Reply:
I see. Thanks!